packages/target-electron/src/deltachat/webxdc.ts

```diff
@@ -89,12 +89,25 @@ export default class DCWebxdc extends SplitOut {
         }
       })
     })
+
+    /**
+     * ipcMain handler for 'open-webxdc' event invoked by the renderer
+     */
     const openWebxdc = async (
       _ev: IpcMainInvokeEvent,
       msg_id: number,
       p: DcOpenWebxdcParameters
     ) => {
-      const { webxdcInfo, chatName, displayname, addr, accountId } = p
+      const { webxdcInfo, chatName, displayname, accountId, href } = p
+      const addr = webxdcInfo.selfAddr || ''
+      let locationHref = ''
+      if (href && href !== '') {
+        const url = new URL(href, 'http://localhost')
+        // make sure url is relative and eval safe
+        locationHref = Buffer.from(
+          url.toString().replace(url.origin, '')
+        ).toString('base64')
```

Comment:
This looks fragile. I feel like it's better to manually construct the final URL.

As an example, `new URL('mailto:example@example.com', 'https://dummy.com').origin === 'null'`.

Am I understanding it correctly that `message.webxdcHref` is user-controlled, that is any user of the app, and the app itself, is free to set it to whatever string they want?
So, clicking a webxdc info message is equivalent to clicking a regular link, except that we have to sanitize the origin of the final URL?
What do you think of doing exactly that? Convert the URLs to absolute URLs and ensure that the origin is the same.

Reply:
>As an example, new URL('mailto:example@example.com', 'https://dummy.com').origin === 'null'.

I assume 'mailto:example@example.com' is an example for a passed webxdc_href
this condition is always false, origin is never null - rather the URL constructor will throw an error. We could use that, to make sure  the passed string is relative, since it throws an error if the string is relative and no base is provided. https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

`url (in the case of absolute URLs) or base + url (in the case of relative references) is not a valid URL.`

But I consider the implementation sufficient: if the string contains  a absolute url, that will be the origin, if not, the provided base is the origin, but in any case we replace the origin with an empty string.

And not to forget, the app itself can set the locationUrl directly anyway, so I consider this change not as critical as displaying a link in a message, for example. And it is a SystemMessage, so can not be easily faked. The href can only be transferred from inside a shared webxdc app.

So if an app takes the hash of a webxdc_href and uses it to "generate" any url in the app running on other chat members device, I could also do whatever I want.

Reply:
> I assume 'mailto:example@example.com' is an example for a passed webxdc_href

Yes.

> this condition is always false, origin is never null

It is `'null'` for the provided example, try putting it in the console. And so `.replace(url.origin, '')` will do nothing.

There are a lot of wonky URLs. `file:///ada/Analytical%20Engine/README.md`, `urn:isbn:9780307476463`. For these `origin` is also `'null'`.

Reply:
> And not to forget, the app itself can set the locationUrl directly anyway

This is not gonna be super on-topic, but we need to make a distinction between the app and a chat member. While an app itself might be secure, chat members can still set this to whatever they want.

Reply:
> So, we don't care about sanitizing the href parameter in any way then, right?

I would argue that we do, for the reason that I mentioned earlier: the app might be secure, but chat members still can't be trusted.
We probably wouldn't want to try to load a URL that has a different origin than the app. Even though a malicious app can still try to do this, IMO we'd want prevent malicious users from doing this.
What do you think?

And sorry for spamming the comments.

Reply:
Let's move this discussion to desktop channel

Reply:
Continued in https://github.com/deltachat/deltachat-desktop/pull/4380#discussion_r1869083111